galley/pkg/config/analysis/analyzers/auth/mtls/policy_checker.go

```diff
@@ -149,72 +182,83 @@ func (pc *PolicyChecker) AddPolicy(namespace string, p *v1alpha1.Policy) error {
-		return true, nil
-	case strictMTLSExplicitlyDisabled:
-		return false, nil
-	case strictMTLSUnset:
-		// Fall through switch case
-	default:
-		return false, fmt.Errorf("unknown strictMTLSMode: %v", pc.serviceToMTLSMode[serviceNoPort])
+	modeAndResource = pc.serviceToMTLSMode[serviceNoPort]
+	if modeAndResource.Resource != nil {
+		return modeAndResource, nil
 	}
 
 	// Check if enforced on namespace
 	namespace, _ := util.GetResourceNameFromHost("", w.FQDN()).InterpretAsNamespaceAndName()
 	if namespace == "" {
-		return false, fmt.Errorf("unable to extract namespace from fqdn: %s", w.FQDN())
+		return ModeAndResource{}, fmt.Errorf("unable to extract namespace from fqdn: %s", w.FQDN())
 	}
-	switch pc.namespaceToMTLSMode[namespace] {
-	case strictMTLSExplicitlyEnabled:
-		return true, nil
-	case strictMTLSExplicitlyDisabled:
-		return false, nil
-	case strictMTLSUnset:
-		// Fall through switch case
-	default:
-		return false, fmt.Errorf("unknown strictMTLSMode: %v", pc.namespaceToMTLSMode[namespace])
+	modeAndResource = pc.namespaceToMTLSMode[namespace]
+	if modeAndResource.Resource != nil {
+		return modeAndResource, nil
 	}
 
-	// Finally, defer to mesh level policy
-	return pc.meshHasStrictMTLSPolicy, nil
+	// Finally, defer to mesh level policy. No need to check for a nil resource
+	// since the default value is correct.
+	return pc.meshMTLSModeAndResource, nil
 }
 
-// doesPolicyEnforceMTLS is a helper function to determine whether or not a
-// Policy implies Strict MTLS mode.
-func doesPolicyEnforceMTLS(p *v1alpha1.Policy) bool {
-	if p.PeerIsOptional {
-		// Connection can still occur.
-		return false
+// TargetServices returns the list of services known to the policy checker (in
+// no particular order).
+func (pc *PolicyChecker) TargetServices() []TargetService {
+	tss := make([]TargetService, len(pc.serviceToMTLSMode))
+	i := 0
+	for ts := range pc.serviceToMTLSMode {
+		tss[i] = ts
+		i++
 	}
+
+	return tss
+}
+
+// parsePolicyMTLSMode is a helper function to determine what mtls mode a Policy
+// implies.
+func parsePolicyMTLSMode(p *v1alpha1.Policy) (Mode, error) {
 	for _, peer := range p.Peers {
 		mtlsParams, ok := peer.Params.(*v1alpha1.PeerAuthenticationMethod_Mtls)
 		if !ok {
 			// Only looking for mtls methods
 			continue
 		}
 
-		// The default value if no Mtls is specified on mtlsParams is strict.
-		// If we do have parameters, though, ensure they do not imply permissive mode.
-		return mtlsParams.Mtls == nil ||
-			(!mtlsParams.Mtls.AllowTls && mtlsParams.Mtls.Mode != v1alpha1.MutualTls_PERMISSIVE)
+		var mode Mode
+		if mtlsParams.Mtls == nil {
+			mode = ModeStrict
+		} else {
+			switch mtlsParams.Mtls.GetMode() {
+			case v1alpha1.MutualTls_PERMISSIVE:
+				mode = ModePermissive
+			case v1alpha1.MutualTls_STRICT:
+				mode = ModeStrict
+			default:
+				// Shouldn't happen!
+				return mode, fmt.Errorf("Unknown MTLS mode when analyzing policy: %s", mtlsParams.Mtls.GetMode().String())
```

Comment:
could happen, right? @diemtvu

Reply:
It could, but not as of this version, right?

Regardless, if it does happen, we handle it by printing an analyzer error, which seems like the most correct thing to do. If a new mode is added that we don't recognize, I'm not sure if we could do any other default behavior correctly.